File: app/src/main/java/com/klim/us_stock/data/retrofit/RetrofitProvider.java
package com.klim.us_stock.data.retrofit;

import androidx.annotation.NonNull;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.klim.us_stock.BuildConfig;

import java.io.IOException;

import okhttp3.HttpUrl;
import okhttp3.Interceptor;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;
import okhttp3.logging.HttpLoggingInterceptor;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

public class RetrofitProvider {
    public static final String BASE_URL = "https://api.polygon.io/";
    public static final String API_KEY = "pBTRT6y6rRicla11X0dQ4V3lRVg1O9cP";
    public static final String REQUEST_DATE_FORMAT = "yyyy-MM-dd";
    public static final String SERVER_TIMEZONE = "GMT-5:00";

    public static Retrofit getRetrofit() {
        HttpLoggingInterceptor interceptor = new HttpLoggingInterceptor();
        interceptor.setLevel(BuildConfig.DEBUG ? HttpLoggingInterceptor.Level.BODY : HttpLoggingInterceptor.Level.NONE);

        Interceptor apiKeyInterceptor = chain -> {
            Request request = chain.request();
            HttpUrl url = request.url().newBuilder().addQueryParameter("apiKey", API_KEY).build();
            request = request.newBuilder().url(url).build();
            return chain.proceed(request);
        };

        OkHttpClient client = new OkHttpClient.Builder()
                .addInterceptor(interceptor)
                .addInterceptor(apiKeyInterceptor)
                .build();

        Gson gson = new GsonBuilder()
                .setLenient()
                .create();

        return new Retrofit.Builder()
                .baseUrl(BASE_URL)
                .addConverterFactory(GsonConverterFactory.create(gson))
                .client(client)
                .build();
    }

}

File: app/src/main/java/com/klim/us_stock/data/retrofit/RetrofitProvider.kt
package com.klim.us_stock.data.retrofit

import com.google.gson.GsonBuilder
import com.klim.us_stock.BuildConfig
import okhttp3.Interceptor
import okhttp3.OkHttpClient
import okhttp3.logging.HttpLoggingInterceptor
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory

object RetrofitProvider {
    const val BASE_URL = "https://api.polygon.io/"
    const val API_KEY = "pBTRT6y6rRicla11X0dQ4V3lRVg1O9cP"
    const val REQUEST_DATE_FORMAT = "yyyy-MM-dd"
    const val SERVER_TIMEZONE = "GMT-5:00"

    val retrofit: Retrofit
        get() = Retrofit.Builder()
            .baseUrl(BASE_URL)
            .addConverterFactory(converter)
            .client(httpClient)
            .build()

    private val converter: GsonConverterFactory
        get() {
            val gson = GsonBuilder()
                .setLenient()
                .create()
            return GsonConverterFactory.create(gson)
        }

    private val httpClient: OkHttpClient
        get() = OkHttpClient.Builder()
            .addInterceptor(loggingInterceptor)
            .addInterceptor(apiKeyInterceptor)
            .build()

    private val loggingInterceptor: Interceptor
        get() {
            val interceptor = HttpLoggingInterceptor()
            interceptor.setLevel(if (BuildConfig.DEBUG) HttpLoggingInterceptor.Level.BODY else HttpLoggingInterceptor.Level.NONE)
            return interceptor
        }

    private val apiKeyInterceptor: Interceptor
        get() = Interceptor { chain: Interceptor.Chain ->
            var request = chain.request()
            val url = request.url.newBuilder().addQueryParameter("apiKey", API_KEY).build()
            request = request.newBuilder().url(url).build()
            chain.proceed(request)
        }
}